org.adempiere.base/src/org/compiere/model/MInOut.java

```diff
@@ -3301,4 +3330,64 @@ public void updateFrom(MOrder order, MInvoice invoice, MRMA rma)
 		}
 		saveEx();
 	}
+	
+	/**
+	 * Stock coverage check - A reversal should not be possible if there is insufficient stock coverage
+	 * @return false when there is insufficient stock coverage
+	 */
+	private boolean stockCoverageCheckForReversal()
+	{
+		MAcctSchema as = MClient.get(getCtx(), Env.getAD_Client_ID(getCtx())).getAcctSchema();
+		if (!MAcctSchema.COSTINGMETHOD_AveragePO.equals(as.getCostingMethod()) 
+				&& !MAcctSchema.COSTINGMETHOD_AverageInvoice.equals(as.getCostingMethod()))
+			return true;
+		
+		final StringBuilder whereClause = new StringBuilder();
+		whereClause.append("AD_Client_ID=? ");
+		whereClause.append("AND C_AcctSchema_ID=? ");
+		whereClause.append("AND M_Product_ID=? ");
+		whereClause.append("AND (DateAcct, M_CostDetail_ID) > ("); 
+		whereClause.append(" SELECT DateAcct, M_CostDetail_ID ");
+		whereClause.append(" FROM M_CostDetail ");
+		whereClause.append(" WHERE M_CostDetail_ID=? ");
+		whereClause.append(") "); 
+		whereClause.append("AND DateAcct >= ? ");
+		whereClause.append("AND Processed='Y' ");
+		whereClause.append("AND (M_InOutLine_ID <> 0 OR C_ProjectIssue_ID <> 0) ");
```

Comment:
why are we checking M_CostDetail here ? Usually, for insufficient storage, we just check M_Cost.

Reply:
Let's say we have a MR (Account Date=01/07/24) with Qty=10, then a shipment (Account Date=02/07/24) with Qty=5. A reversal of shipment is required to 'replenish' stock, prior to reversal of the MR (Account Date=01/07/24).